### PR TITLE
Price feed uses rolling average

### PR DESCRIPTION
**Pull request by thodges-gh**
Finishes [#164214424](https://www.pivotaltracker.com/story/show/164214424).

The story only requires 3 oracles, but I found it just as easy to build it without that constraint, and tests up to 5 (and it's pretty clear how it would work with more). I also added the tslint script to truffle_test since it's really useful and won't hurt anything. One thing I think we might consider changing is using the block's timestamp instead of its number for averaging recent updates. Truffle starts the block number at 0, so we have to mine some empty blocks first in order for the tests to run.

### CODE REVIEW

File: solidity/contracts/LinkEx.sol [hunk 1]
```diff
@@ -18,15 +37,46 @@ contract LinkEx is LinkExInterface {
...
+
+  function update(uint256 _rate) external onlyAuthorizedNode {
+    rates[msg.sender] = Rate(block.number, _rate);
     if (isFutureBlock()) {
       historicRate = rate;
       rateHeight = block.number;
     }
-    rate = _rate;
+
+    uint256 tempRate;
```
Comment (j16r) [inline solidity/contracts/LinkEx.sol]: Something like `rateSum` might be a better name.

File: solidity/build/contracts/Oracle.json [hunk 2]
```diff
@@ -355,84 +355,84 @@
       "type": "function"
```
Comment (j16r) [inline solidity/build/contracts/Oracle.json]: Ooops, we should .gitignore and rm this.

File: solidity/test/LinkEx_test.ts [hunk 3]
```diff
@@ -1,24 +1,195 @@
...
+
+        it('the removed oracles do not contribute to the average', async () => {
+          const rate = await contract.currentRate()
+          assert.equal(rate, newExpectedAvg)
+        })
+      })
+
+      context('when updated by an oracle after 25 blocks', () => {
+        beforeEach(async () => {
+          for (let i = 0; i < 25; i++) {
```
Comment (j16r) [inline solidity/test/LinkEx_test.ts]: Since we do this in multiple paces might be best to wrap as a helper e.g: `h.mineBlocks(25)` ?

File: solidity/contracts/LinkEx.sol [hunk 4]
```diff
@@ -1,15 +1,34 @@
...
+
+  uint256 public constant VALID_BLOCKS = 25;
+
+  mapping(address => bool) public authorizedNodes;
+
+  mapping(address => Rate) private rates;
   uint256 private historicRate;
   uint256 private rate;
   uint256 private rateHeight;
+  address[] private oracles;
```
Comment (j16r) [inline solidity/contracts/LinkEx.sol]: Do we need this *and* authorizedNodes above?

File: solidity/build/contracts/Oracle.json (hunk 2 above)
Comment (thodges-gh) [reply solidity/build/contracts/Oracle.json]: This was added for https://github.com/smartcontractkit/chainlink/pull/950.

File: solidity/contracts/LinkEx.sol (hunk 4 above)
Comment (thodges-gh) [reply solidity/contracts/LinkEx.sol]: `authorizedNodes` is a mapping, and Solidity doesn't let you iterate over mappings.

File: solidity/contracts/LinkEx.sol [hunk 5]
```diff
@@ -18,15 +37,46 @@ contract LinkEx is LinkExInterface {
...
   }
 
-  function update(uint256 _rate) external {
+  function removeOracle(address _oracle) external onlyOwner {
+    setFulfillmentPermission(_oracle, false);
+    delete authorizedNodes[_oracle];
+    for (uint i = 0; i < oracles.length; i++) {
+      if (oracles[i] == _oracle) {
+        delete rates[oracles[i]];
+        delete oracles[i];
```
Comment (se3000) [inline solidity/contracts/LinkEx.sol]: Does this reorder the list? I think it just leaves an empty element in that index, so when we decrement the length in the next line, it leaves an empty spot. If I'm right, it would be exposed with a test involving three oracles, and oracle number two gets deleted.

One solution might be to get the last oracle in the list, delete it from there and copy it to the previous location, but that may well have some edge cases to it.

File: solidity/test/LinkEx_test.ts [hunk 6]
```diff
@@ -1,24 +1,191 @@
-import { bigNum, deploy } from './support/helpers'
+import * as h from './support/helpers'
 
 contract('LinkEx', () => {
   let contract: any
 
+  // Need to mine some blocks so that the check in update doesn't
+  // fail when subtracting 25 from block.number.
+  before(async () => {
+    h.mineBlocks(50)
```
Comment (se3000) [inline solidity/test/LinkEx_test.ts]: Why not make it 25 if that's the number we care about?

File: solidity/test/LinkEx_test.ts [hunk 7]
```diff
@@ -1,24 +1,191 @@
...
+    context('when requested in the same block as an update', () => {
+      const expected = 3542157117
+      const updated = 8616460799
+
+      beforeEach(async () => {
+        await contract.addOracle(h.oracleNode, {from: h.defaultAccount})
+        await contract.update(expected, {from: h.oracleNode})
+      })
+
+      it('returns the historic rate', async () => {
```
Comment (se3000) [inline solidity/test/LinkEx_test.ts]: 👍 Cool test.

File: solidity/contracts/LinkEx.sol (hunk 5 above)
Comment (thodges-gh) [reply solidity/contracts/LinkEx.sol]: Okay good. We shouldn't be touching the length because that will effect every element of the array after it. The test was only deleting the last element so there didn't appear to be a problem, but when updating that to delete the first element is where everything goes wrong. This is now corrected by simply leaving the length as-is on removal of an oracle.

Comment (se3000) [reply solidity/contracts/LinkEx.sol]: I don't think it's that simple, we now have an empty oracle in the list, ie `oracles[i] == adress(0)`, right? I think we need to decrement the count. Then, reorder the oracle array, except if the oracle deleted is the last oracle in the list.

Comment (thodges-gh) [reply solidity/contracts/LinkEx.sol]: Turns out it was simple, just not as efficient as it could be. We did have a test already for deleting an oracle at the beginning of the array and running the average again. What happened was the update loop would hit that empty place in the array, see that blocknumber (being 0) was older than 25 blocks, and not count it towards the average. Now upon removing an oracle, if the target oracle isn't the last in the array, the last oracle in the array will take the place of the deleted oracle, and we'll decrement the array length. This seems to save ~1000 gas per-update in my small-scale test.

Comment (se3000) [reply solidity/contracts/LinkEx.sol]: We'd also get a gas refund if we also deleted the last spot in that array, right?

**Comment by se3000** [comment]
Also, can an Oracle be permissioned multiple times? If so, does that allow them to respond multiple times? Does it change any of our assumptions?

File: solidity/contracts/LinkEx.sol (hunk 5 above)
Comment (thodges-gh) [reply solidity/contracts/LinkEx.sol]: I had to refactor that `removeOracle` method a bit to get some more gas savings out of it, but yes:
Start when deleting from middle of array: 42324
After adding the line to delete the last oracle element: 45242
After adding the line to calculate size once: 44940
After adding return values to end early once removed: 44434
After removing the duplicate setFulfillmentPermission method call: 41738
After using the calculated size for the new length: 41584

**Comment by thodges-gh** [comment]
> Also, can an Oracle be permissioned multiple times? If so, does that allow them to respond multiple times? Does it change any of our assumptions?

This was an easy fix to implement, simply require that the node is not authorized at the beginning of `addOracle`, and I added a test for it.

**Comment by thodges-gh** [comment]
One security concern that I thought of: if an oracle feeds bad data and we need to remove it, their malicious data will remain a part of the average until another oracle updates. The fix for this was to refactor the `update` method a bit so that it can be called externally by an oracle with an updated price, but that calls a private `_update` method which simply calculates the stored prices. When an oracle is removed, we'll call that private `_update` method again to create a new average on the same block.

File: solidity/contracts/LinkEx.sol (hunk 5 above)
Comment (thodges-gh) [reply solidity/contracts/LinkEx.sol]: Actually I just came across [this](https://ethereum.stackexchange.com/a/39311/37908) answer. I'll do some tests to see if this holds true. If so, we might be able to save another ~3000 or so gas when removing an oracle.

Comment (se3000) [reply solidity/contracts/LinkEx.sol]: Based on that SO answer, can't we get rid of the `delete oracles[i];` on line 53? Seems like it is handled by 54-57.

File: solidity/contracts/LinkEx.sol [hunk 8]
```diff
@@ -18,15 +37,68 @@ contract LinkEx is LinkExInterface {
...
 
-  function update(uint256 _rate) external {
+  function removeOracle(address _oracle) external onlyOwner {
+    _removeOracle(_oracle);
+    if (oracles.length > 0) {
+      _update();
+    }
+  }
+
+  function _removeOracle(address _oracle) private {
```
Comment (se3000) [inline solidity/contracts/LinkEx.sol]: I'd prefer to be verbose and have explicit names, instead of denoting private methods(introduced purely for gas savings) with an underscore. Can we call the private method something like `deauthorizeAddress`?
